File: backend/app/schemas/setting.py
"""系统设置（LLM 模型配置）的请求 / 响应模型。"""
from pydantic import BaseModel, ConfigDict, Field, field_validator
from pydantic.alias_generators import to_camel

_CFG = ConfigDict(alias_generator=to_camel, populate_by_name=True)


def _check_base_url(v: str) -> str:
    """供应商地址必须是 http(s) 绝对地址，避免存进一个没法请求的值。"""
    url = v.strip()
    if not url.startswith(("http://", "https://")):
        raise ValueError("API 地址需以 http:// 或 https:// 开头")
    return url


class LLMSettingOut(BaseModel):
    """设置页回显：包含脱敏后的 Key，绝不回传明文 Key。"""

    model_config = _CFG

    enabled: bool = False
    # 是否已配置 API Key
    has_api_key: bool = False
    # 脱敏预览，如 sk-****abcd
    api_key_preview: str = ""
    base_url: str = ""
    model: str = ""
    timeout_seconds: float = 30.0
    min_change_lines: int = 3
    # real = 走真实模型；mock = 规则兜底
    mode: str = "mock"
    # database = 来自设置页保存；env = 尚未保存，读的是 .env 默认值
    source: str = "env"
    # 一句话状态说明，前端直接展示
    message: str = ""


class LLMSettingUpdate(BaseModel):
    """保存设置。

    api_key 的语义：
    - 省略（None）→ 保持不变
    - 空串 ""    → 清除已保存的 Key
    - 有值       → 覆盖
    clear_api_key=true 时无条件清除（前端点"清除 Key"时用）。
    """

    model_config = _CFG

    enabled: bool | None = None
    api_key: str | None = None
    clear_api_key: bool = False
    base_url: str | None = None
    model: str | None = None
    timeout_seconds: float | None = Field(default=None, ge=3, le=300)
    min_change_lines: int | None = Field(default=None, ge=1, le=100)


class LLMProviderOut(BaseModel):
    """供应商列表项：Key 只回传脱敏预览。"""

    model_config = _CFG

    id: int
    name: str = ""
    base_url: str = ""
    model: str = ""
    has_api_key: bool = False
    api_key_preview: str = ""
    # 运行时正在使用这一条
    is_active: bool = False
    # 已缓存的模型数量。完整列表走 /llm/providers/{id}/models 按需取，
    # 避免大列表在每次列表请求里全量传输
    models_count: int = 0


class LLMModelsOut(BaseModel):
    """某条供应商已缓存的完整模型列表（编辑时按需获取）。"""

    model_config = _CFG

    models: list[str] = Field(default_factory=list)


class LLMProviderCreate(BaseModel):
    """添加供应商：地址 / 模型必填，Key 必填（否则无法走真实模型）。"""

    model_config = _CFG

    name: str | None = Field(default=None, max_length=80)
    base_url: str = Field(..., min_length=1, max_length=255)
    model: str = Field(..., min_length=1, max_length=120)
    api_key: str = Field(..., min_length=1, max_length=500)
    # 添加前已从上游拉到的模型列表（可选）
    models: list[str] | None = None

    @field_validator("base_url")
    @classmethod
    def _validate_base_url(cls, v: str) -> str:
        return _check_base_url(v)


class LLMProviderUpdate(BaseModel):
    """编辑供应商：字段均可省略只改传了的；api_key 语义同 LLMSettingUpdate。"""

    model_config = _CFG

    name: str | None = Field(default=None, max_length=80)
    base_url: str | None = Field(default=None, min_length=1, max_length=255)
    model: str | None = Field(default=None, min_length=1, max_length=120)
    api_key: str | None = Field(default=None, min_length=1, max_length=500)
    clear_api_key: bool = False
    # 省略 = 保持不变；传了就覆盖（用于保存本次新拉到的模型列表）
    models: list[str] | None = None

    @field_validator("base_url")
    @classmethod
    def _validate_base_url(cls, v: str | None) -> str | None:
        return _check_base_url(v) if v is not None else None


class LLMProviderReorder(BaseModel):
    """拖动排序：按期望顺序传一组供应商 id。"""

    model_config = _CFG

    ids: list[int] = Field(..., min_length=1)


class LLMTestRequest(BaseModel):
    """测试连接：优先用请求里的值，缺省则用已保存/环境里的值。

    provider_id：测试「已保存的供应商」时传入——它的 Key 前端拿不到（脱敏），
    由后端自己取已存 Key 来测。
    """

    model_config = _CFG

    provider_id: int | None = None
    base_url: str | None = None
    model: str | None = None
    api_key: str | None = None
    timeout_seconds: float | None = Field(default=None, ge=3, le=300)


class LLMTestResult(BaseModel):
    model_config = _CFG

    ok: bool
    message: str
    model: str = ""
    latency_ms: int = 0


class LLMApiKeyOut(BaseModel):
    """已保存 API Key 的明文回显：仅登录用户、且仅在编辑某条供应商时按需获取。"""

    model_config = _CFG

    api_key: str = ""


class LLMFetchModelsRequest(BaseModel):
    """从上游供应商拉取模型列表：优先用已保存供应商的地址/Key（前端拿不到明文）。"""

    model_config = _CFG

    provider_id: int | None = None
    base_url: str | None = None
    api_key: str | None = None
    timeout_seconds: float | None = Field(default=None, ge=3, le=300)


class LLMFetchModelsResult(BaseModel):
    """拉取结果：ok=false 时 models 为空，message 给可读原因。"""

    model_config = _CFG

    ok: bool = False
    models: list[str] = Field(default_factory=list)
    message: str = ""

File: frontend/mock/setting.ts
import { authUser, db, err, idFrom, ok } from "./db";

function requireUser(headers: any) {
  return authUser(headers) || null;
}

// ---------- 序列化 ----------
function previewOf(apiKey: string | undefined): string {
  if (!apiKey) return "";
  return apiKey.length <= 4 ? "****" : `sk-****${apiKey.slice(-4)}`;
}

function providerOut(p: any) {
  return {
    id: p.id,
    name: p.name || "",
    baseUrl: p.baseUrl,
    model: p.model,
    hasApiKey: !!p.hasApiKey,
    apiKeyPreview: p.hasApiKey ? previewOf(p.apiKey) : "",
    isActive: !!p.isActive,
    modelsCount: (p.models || []).length,
  };
}

// ---------- 当前用户的 LLM 数据（与后端按 user_id 隔离一致） ----------
function ownProviders(d: any, uid: number): any[] {
  return d.providers.filter((p: any) => p.userId === uid);
}
function activeProvider(d: any, uid: number): any | null {
  return ownProviders(d, uid).find((p: any) => p.isActive) || null;
}
function getSetting(d: any, uid: number): any {
  let s = d.llmSetting[uid];
  if (!s) {
    s = { enabled: false, timeoutSeconds: 30, minChangeLines: 3, source: "env" };
    d.llmSetting[uid] = s;
  }
  return s;
}
function findOwnedProvider(d: any, uid: number, id: number): any | null {
  return ownProviders(d, uid).find((p: any) => p.id === id) || null;
}
function maxSortOrder(d: any, uid: number): number {
  return ownProviders(d, uid).reduce((m, p) => Math.max(m, p.sortOrder ?? 0), 0);
}

// 开关 + 使用中的供应商 → 展示配置（对齐后端 _compose + _to_out）
function llmOut(d: any, uid: number) {
  const s = getSetting(d, uid);
  const prov = activeProvider(d, uid);
  const apiKey = prov ? prov.apiKey || "" : "";
  const baseUrl = prov ? prov.baseUrl || "" : "";
  const model = prov ? prov.model || "" : "";
  const hasApiKey = !!apiKey;
  const mode = apiKey && baseUrl && model ? "real" : "mock";
  const source = s.source === "database" || prov ? "database" : "env";
  let message: string;
  if (mode === "real") message = `已启用真实模型（${model}）`;
  else if (s.enabled && !apiKey) message = "已开启 AI 分析，但尚未填写 API Key，当前仍走规则 Mock";
  else if (s.enabled) message = "已开启 AI 分析，但配置不完整（缺少地址或模型），当前仍走规则 Mock";
  else message = "未启用 AI 分析，当前走规则 Mock（无需 Key 也能跑通全链路）";
  return {
    enabled: !!s.enabled,
    hasApiKey,
    apiKeyPreview: hasApiKey ? previewOf(apiKey) : "",
    baseUrl,
    model,
    timeoutSeconds: s.timeoutSeconds ?? 30,
    minChangeLines: s.minChangeLines ?? 3,
    mode,
    source,
    message,
  };
}

function statusOut(d: any, uid: number) {
  const l = llmOut(d, uid);
  return {
    mode: l.mode,
    enabled: l.enabled,
    hasApiKey: l.hasApiKey,
    baseUrl: l.baseUrl,
    model: l.model,
    timeoutSeconds: l.timeoutSeconds,
    minChangeLines: l.minChangeLines,
    message:
      l.mode === "real"
        ? `已启用真实模型（${l.model} @ ${l.baseUrl}）`
        : "未配置 Key，使用规则 Mock 兜底（可在「设置」页配置模型）",
  };
}

export default [
  // 静态路由必须排在 :id 之前
  {
    url: "/api/settings/llm",
    method: "get",
    timeout: 200,
    response: ({ headers }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      return ok(llmOut(db(), u.id));
    },
  },
  {
    url: "/api/settings/llm",
    method: "put",
    timeout: 200,
    response: ({ headers, body }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const s = getSetting(d, u.id);
      const p = body || {};
      if (p.enabled !== undefined) s.enabled = !!p.enabled;
      if (p.timeoutSeconds !== undefined) s.timeoutSeconds = Number(p.timeoutSeconds) || 30;
      if (p.minChangeLines !== undefined) s.minChangeLines = Number(p.minChangeLines) || 3;
      s.source = "database";
      return ok(llmOut(d, u.id));
    },
  },
  {
    url: "/api/settings/llm/test",
    method: "post",
    timeout: 300,
    response: ({ headers, body }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const id = body?.providerId ? Number(body.providerId) : null;
      const prov = id != null ? findOwnedProvider(d, u.id, id) : activeProvider(d, u.id);
      const model = body?.model || prov?.model || "gpt-4o";
      return ok({ ok: true, message: "连接成功，模型可用", model, latencyMs: 120 });
    },
  },
  {
    url: "/api/settings/llm/providers",
    method: "get",
    timeout: 200,
    response: ({ headers }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const list = ownProviders(d, u.id).sort((a, b) => (a.sortOrder ?? 0) - (b.sortOrder ?? 0));
      return ok(list.map(providerOut));
    },
  },
  {
    url: "/api/settings/llm/providers",
    method: "post",
    timeout: 300,
    response: ({ headers, body }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const baseUrl = (body?.baseUrl || "").trim();
      if (!baseUrl) return err(40002, "baseUrl 不能为空");
      const own = ownProviders(d, u.id);
      const isFirst = own.length === 0;
      const row = {
        id: d.seq.provider++,
        userId: u.id,
        name: body?.name || "",
        baseUrl,
        model: body?.model || "",
        apiKey: body?.apiKey || "",
        hasApiKey: !!(body?.apiKey || (body?.providerId ? true : false)),
        isActive: isFirst || !!body?.isActive,
        sortOrder: maxSortOrder(d, u.id) + 1,
        models: Array.isArray(body?.models) ? body.models : [],
      };
      if (row.isActive) own.forEach((p: any) => (p.isActive = false));
      d.providers.push(row);
      return ok(providerOut(row));
    },
  },
  {
    url: "/api/settings/llm/providers/reorder",
    method: "post",
    timeout: 200,
    response: ({ headers, body }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const ids: number[] = Array.isArray(body?.ids) ? body.ids.map(Number) : [];
      const own = ownProviders(d, u.id);
      const map = new Map(own.map((p: any) => [p.id, p]));
      if (ids.some((id) => !map.has(id))) return err(40405, "供应商不存在或已被删除");
      ids.forEach((id, i) => (map.get(id).sortOrder = i));
      return ok(null);
    },
  },
  {
    url: "/api/settings/llm/providers/models",
    method: "post",
    timeout: 300,
    response: ({ headers }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      return ok({ ok: true, models: ["deepseek-chat", "deepseek-reasoner", "gpt-4o-mini", "gpt-4o"], message: "已获取模型列表" });
    },
  },
  {
    url: "/api/llm/status",
    method: "get",
    timeout: 200,
    response: ({ headers }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      return ok(statusOut(db(), u.id));
    },
  },
  {
    url: "/api/settings/llm/providers/:id",
    method: "put",
    timeout: 300,
    response: ({ headers, url, body }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const id = idFrom(url);
      const p = findOwnedProvider(d, u.id, id);
      if (!p) return err(40405, "供应商不存在");
      const patch = body || {};
      if (patch.name !== undefined) p.name = patch.name ?? "";
      if (patch.baseUrl !== undefined) p.baseUrl = patch.baseUrl;
      if (patch.model !== undefined) p.model = patch.model;
      if (patch.clearApiKey) {
        p.apiKey = "";
        p.hasApiKey = false;
      } else if (patch.apiKey !== undefined && patch.apiKey !== "") {
        p.apiKey = patch.apiKey;
        p.hasApiKey = true;
      }
      if (Array.isArray(patch.models)) p.models = patch.models;
      return ok(providerOut(p));
    },
  },
  {
    url: "/api/settings/llm/providers/:id",
    method: "delete",
    timeout: 200,
    response: ({ headers, url }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const id = idFrom(url);
      const own = ownProviders(d, u.id);
      const idx = own.findIndex((x: any) => x.id === id);
      if (idx < 0) return err(40405, "供应商不存在");
      const wasActive = own[idx].isActive;
      d.providers = d.providers.filter((x: any) => !(x.id === id && x.userId === u.id));
      if (wasActive) {
        const rest = ownProviders(d, u.id).sort((a: any, b: any) => b.id - a.id);
        if (rest.length) rest[0].isActive = true;
      }
      return ok(null);
    },
  },
  {
    url: "/api/settings/llm/providers/:id/activate",
    method: "post",
    timeout: 200,
    response: ({ headers, url }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const id = idFrom(url);
      const p = findOwnedProvider(d, u.id, id);
      if (!p) return err(40405, "供应商不存在");
      ownProviders(d, u.id).forEach((x: any) => (x.isActive = x.id === id));
      return ok(providerOut(p));
    },
  },
  {
    url: "/api/settings/llm/providers/:id/duplicate",
    method: "post",
    timeout: 200,
    response: ({ headers, url }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const id = idFrom(url);
      const p = findOwnedProvider(d, u.id, id);
      if (!p) return err(40405, "供应商不存在");
      const copy = JSON.parse(JSON.stringify(p));
      copy.id = d.seq.provider++;
      copy.name = copy.name ? `${copy.name} 副本` : "";
      copy.isActive = false;
      copy.sortOrder = maxSortOrder(d, u.id) + 1;
      d.providers.push(copy);
      return ok(providerOut(copy));
    },
  },
  {
    url: "/api/settings/llm/providers/:id/api-key",
    method: "get",
    timeout: 200,
    response: ({ headers, url }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const id = idFrom(url);
      const p = findOwnedProvider(d, u.id, id);
      if (!p) return err(40405, "供应商不存在");
      return ok({ apiKey: p.apiKey || "" });
    },
  },
  {
    url: "/api/settings/llm/providers/:id/models",
    method: "get",
    timeout: 200,
    response: ({ headers, url }: any) => {
      const u = requireUser(headers);
      if (!u) return err(40100, "未登录或登录已过期");
      const d = db();
      const id = idFrom(url);
      const p = findOwnedProvider(d, u.id, id);
      if (!p) return err(40405, "供应商不存在");
      return ok({ models: p.models || [] });
    },
  },
];